app/controllers/registration/registration.controller.js

```diff
@@ -102,15 +106,55 @@ async function showAuthenticatorAppPage (req, res, next) {
     const otpUrl = `otpauth://totp/GOV.UK%20Pay:${encodeURIComponent('username')}?secret=${encodeURIComponent(secretKey)}&issuer=GOV.UK%20Pay&algorithm=SHA1&digits=6&period=30`
     const qrCodeDataUrl = await qrcode.toDataURL(otpUrl)
 
+    const recovered = sessionData.recovered || {}
+    delete sessionData.recovered
+
     res.render('registration/authenticator-app', {
       prettyPrintedSecret,
-      qrCodeDataUrl
+      qrCodeDataUrl,
+      errors: recovered.errors
     })
   } catch (err) {
     next(err)
   }
 }
 
+async function submitAuthenticatorAppPage (req, res, next) {
+  const sessionData = req.register_invite
+  if (!registrationSessionPresent(sessionData)) {
+    return next(new RegistrationSessionMissingError())
+  }
+
+  const otpCode = sanitiseSecurityCode(req.body[OTP_CODE_FIELD_NAME])
+  const validationResult = validateOtp(otpCode)
+
+  const errors = {}
+  if (!validationResult.valid) {
+    errors[OTP_CODE_FIELD_NAME] = validationResult.message
+    sessionData.recovered = { errors }
+    return res.redirect(paths.register.authenticatorApp)
+  }
+
+  try {
+    await adminusersClient.verifyOtpForInvite(sessionData.code, otpCode)
+    const completeInviteResponse = await adminusersClient.completeInvite(sessionData.code)
+    const userExternalId = completeInviteResponse.user_external_id
+    loginController.setupDirectLoginAfterRegister(req, res, userExternalId)
```

Comment:
Any reason controller is calling another controller? Can it be in a service method ?

Looks like the only statement required from `setupDirectLoginAfterRegister` is `req.register_invite.userExternalId = userExternalId` as userExternalId will never be null when calling from `registration.controller.js`

Reply:
Yes I copied this code from the existing controllers and was meaning to change this in another PR, but actually probably makes sense to change that as part of this PR so its not forgotten about. Will make the change.

Reply:
updated